### ISSUE TITLE
service metrics: extend to non-transaction events and add zero counts

### ISSUE THREAD

**Issue by simitt** [elastic/apm-server#9654]
Service metrics should also be produced when application metrics, errors & logs for a service are observed. When during an aggregation period no transaction events are seen for this service, then a zero count for transaction throughput needs to be stored with the service metric.

**Comment by simitt** [comment]
Any non trace event should trigger the creation of a metricset, but with zero values for the `Metricset` and `Transaction` counts. The aggregation logic needs to be changed to not only process transaction events.
What's problematic is that the aggregation key contains the `transaction.type`. The type is not known if no transaction event is available. @felixbarny does it make sense to create the service metrics with `transaction.type: "unknown"` for the UI queries?

**Comment by felixbarny** [comment]
Good question, Silvia. I don't have a clear answer but a few thoughts.

> The aggregation logic needs to be changed to not only process transaction events.

This is a bit at odds with https://github.com/elastic/apm-server/issues/5243#issuecomment-1345504925

> I would strongly prefer that service metrics remain a subset of transaction metrics so that we could later derive them in Elasticsearch by just dropping dimensions.

Maybe we should have another metricset (or metricsets) for different signals that summarize, for example, the rate of metrics, logs, errors, and maybe the number of profiling frames for a service. These metricsets don't have a `transaction.type` dimension. 

Therefore, the closest thing to an answer to your question

> does it make sense to create the service metrics with transaction.type: "unknown" for the UI queries?

is that there shouldn't be a `transaction.type` dimension for non-trace-related service metrics.

Another issue is that not all signals are necessarily passing through APM Server (for example logs that are shipped by Filebeat). So if we only rely on service metrics in the service inventory UI (which is the premise of this issue), we might miss some services. I think it's important to show services that only have logs, for example. We could query the logs index in addition to the service metrics but the risk is that it's going to be very slow.

cc @dgieselaar

**Comment by felixbarny** [comment]
To make a concrete suggestion, APM server could emit a service metric document that only has a `service.name` and a `service.environment` dimension and nothing else if it receives any signal from a service. 

The document may look like this (note the lack of any numerical metric value):

```json
{
  "@timestamp": "2006-01-02T15:04:05.000Z",
  "metricset.name": "service",
  "service": {
    "environment": "prod",
    "name": "systemtest"
  }
}
```

This would be in addition to the service metrics that have a `transaction.duration` metric and a `transaction.type` dimension. Alternatively, we could emit these metrics only if there are no service-transaction metrics.

If it is an issue that these metrics have the same `metricset.name` as the service metrics that have `transaction.duration` information, we probably need to introduce dedicated metricset names, like `service-transaction`, `service-metric`, `service-log`. The UI would then query all of them to create the service inventory table. But for latency and throughput metrics and sparklines, it would only query the `service-transaction` metricset. Still all of these would be stored in the same data stream.

**Comment by felixbarny** [comment]
I want to add that this issue is critical to improve the service inventory latency. Otherwise, the UI needs to query all metrics, including those that are not rolled up (such as JVM metrics) and most of the performance improvements of service metrics and continuous rollups would be moot.

**Comment by dgieselaar** [comment]
I like the idea of creating a separate document (e.g. a heartbeat), but can we make sure we have something other than `metricset.name:service` for these documents?

**Comment by simitt** [comment]
I believe it would be easiest to keep all other dimensions that are reported for service metrics, and at least internally set `transaction.type: "unknown"` an report zero values for the transaction counts (as originally requested). On write we could filter out `transaction.type` if set to `unknown`, so that it doesn't show up in the ES documents.

**Comment by simitt** [comment]
@felixbarny @dgieselaar does the above make sense to you? We need to get to an agreement before the implementation can start.

**Comment by dgieselaar** [comment]
@simitt I'd prefer an empty string or `unknown` so we don't have to use the `missing` option for terms/composite aggs which has worse performance.

**Comment by dgieselaar** [comment]
One thing to keep in mind here is that Elasticsearch does not allow `_doc_count` to be 0. Perhaps we should create an entirely different metric for non-transaction events? (it's not clear to me from this discussion that that is the agreement).

**Comment by dgieselaar** [comment]
@simitt We don't want to query _all_ the metrics to determine what services are active (but not necessarily recording transaction data). For service metrics, `_doc_count` is referring to the amount of transactions, of which there are none in these cases. There's no way to set `_doc_count` to 0, so we cannot use service metrics for this purpose (AFAIK), so our only option is to create an entirely separate document, that aggregates over `service.name` and `service.environment` only. If we query that data in parallel to service metrics, it will ~always be faster, because cardinality is guaranteed to be lower than service metrics.

**Comment by axw** [comment]
> If it is an issue that these metrics have the same metricset.name as the service metrics that have transaction.duration information, we probably need to introduce dedicated metricset names, like service-transaction, service-metric, service-log. The UI would then query all of them to create the service inventory table. But for latency and throughput metrics and sparklines, it would only query the service-transaction metricset. Still all of these would be stored in the same data stream.

I'd be keen to rename `service` metrics to `service_transaction` metrics. It is increasingly clear to me that just "service" is a misnomer - they're measuring a specific aspect of the service.

I'm not entirely sure about having `service_metric`, `service_log`, etc. Maybe `service_summary`? This could count the number of transaction, span, metricset, and log events processed per service. This one `service_summary` metricset/data stream would be used to build the list of services. The `service_transaction` metrics would then be overlaid for services that are processing transactions.

**Comment by dgieselaar** [comment]
@simitt @axw I want to emphasise that at least the schema for these zero-count metrics should be addressed asap if we want to ship a meaningful performance improvement to the service inventory. We cannot start work without it. Without these metrics we still have a blocking request to fetch all services from all metrics, which is a lot slower than the query for service metrics only.

**Comment by felixbarny** [comment]
+1 to Andrew's suggestion. Would that work for you, @dgieselaar?

**Comment by dgieselaar** [comment]
@felixbarny @axw yes. I don't see a use for the counts yet (for `service_summary`, but I'm not against adding them either. I'm also in favor of renaming `service` to `service_transaction` now that we still have the opportunity to do so without a breaking change/bwc gymnastics.

**Comment by axw** [comment]
To keep things simple and get this done ASAP I think we can skip the counts for now. So I propose the following fields:

- `@timestamp`
- `metricset.name: service_summary`
- `metricset.interval`
- `processor.event: metric`
- `processor.name: metric`
- `service.environment`
- `service.name`
- `agent.name`
- `labels.*` (global labels only)
- `numeric_labels.*` (global labels only)

Then we can add counts later if/when we have a use case for them. WDYT?

**Comment by dgieselaar** [comment]
@axw sounds good to me. We should perhaps consider only emitting them when there are zero transactions, but we can do that later. The only difference in terms of cardinality will be transaction.type, right?

**Comment by axw** [comment]
>  The only difference in terms of cardinality will be transaction.type, right?

correct

**Comment by felixbarny** [comment]
If adding counts adds complexity and therefore risk to deliver the feature, we should skip them for now. But it's a little awkward to have no metrics but just metadata.

The counts may look like this:
* `transaction.count`
* `error.count`
* `log.count`
* `metric.count`

We could use them to determine which tabs to show in the service overview.

The thing that makes me feel a bit uneasy about this is that this only considers signals that pass though APM Server. If metrics or logs are ingested directly to ES, they're not considered here which means we won't even list the service. But I don't see a good alternative.

**Comment by carsonip** [comment]
Just wanted to make sure what's needed in the implementation:
1. a new service summary aggregator that emits to a new data stream `service_summary`
2. the data stream has these fields:
> 
    @timestamp
    metricset.name: service_summary
    metricset.interval
    processor.event: metric
    processor.name: metric
    service.environment
    service.name
    agent.name
    labels.* (global labels only)
    numeric_labels.* (global labels only)
3. aggregation key includes `service.environment`, `service.name` only (as mentioned in discussion). `agent.name` and global labels are not included.
4. the aggregator listens to all events with `event.Processor == model.SpanProcessor` or `event.Processor == model.TransactionProcessor`

Question:
1. should this aggregator have intervals 1m,10m,60m like service metrics and transaction metrics?
2. is (3) the proposed aggregation key right? How do we emit `agent.name` and global labels if they are not in the aggregation key?
3. is (4) above about `event.Processor` correct?
4. is it ok to split the work of renaming `service` to `service_transaction` to another PR?

**Comment by felixbarny** [comment]
3. `agent.name` should not be part of the aggregation key, I think
4. I think it doesn't need to listen for spans but it does need to listen for transactions, errors, metrics, and logs

**Comment by axw** [comment]
> 1. should this aggregator have intervals 1m,10m,60m like service metrics and transaction metrics?

Yes

> 2. is (3) the proposed aggregation key right? @dgieselaar mentioned "aggregates over service.name and service.environment only"

Yes. It should be the same as `service_transaction` but without `transaction.type`. Someone correct me if I'm mistaken, but I think we _need_ `agent.name` so we can render the language icon next to the service name in the services inventory:

![image](https://user-images.githubusercontent.com/843579/211971256-1598b816-31a6-4c07-b4d0-9313af434ebe.png)

> 3. is (4) above about event.Processor correct?

The `service_summary` aggregator should be considering all event types. That includes MetricProcessor, LogProcessor, and ErrorProcessor. I think we should just not filter on Processor at all.

> 4. is it ok to split the work of renaming service to service_transaction to another PR?

Yes, perfectly fine

**Comment by carsonip** [comment]
Question about implementation: Should service summary metrics get its own `enabled` config like service metrics? imo it should be enabled when service metrics aggregation is enabled, but it needs a separate MaxGroups config possibly. One possible solution is that service summary metrics does not have its own config and uses service metrics `enabled` and `max_groups`. Any opinion on overloading the service metrics config?

**Comment by simitt** [comment]
>Should service summary metrics get its own enabled config like service metrics? imo it should be enabled when service metrics aggregation is enabled, but the aggregator needs a MaxGroups config. One possible solution is that service summary metrics reuses service metrics enabled and max_groups. Any opinion on overloading the service metrics config?

With https://github.com/elastic/apm-server/issues/10048 we will enable service metrics and service summary metrics by default and can remove the `enabled` configuration option completely. 

Let's reuse the `max_groups` from service metrics for now. We can always split it out into a dedicated config option if needed, but would avoid exposing more options than needed to the user for now.

**Comment by carsonip** [comment]
This task is split into:
- Add service summary metrics aggregator: #10029 
- Add overflow handling to service summary metrics: TODO
- Rename service metrics to service transaction metrics: TODO

**Comment by carsonip** [comment]
How far do we want to go with the renaming? Are we fine with changing data stream name and config field name `apm-server.aggregation.service`?

**Comment by simitt** [comment]
@carsonip can you summarize the test guidance for this issue please
